### PR TITLE
SONARIAC-1405 Introduce `HasIdentifier` node to represent `variables()` and `parameters()` in JSON and identifiers in Bicep

### PR DESCRIPTION
**Pull request by petertrr**

### CODE REVIEW

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/BicepGrammar.java [hunk 1]
```diff
@@ -754,6 +756,12 @@ public Identifier IDENTIFIER() {
         b.token(BicepLexicalGrammar.IDENTIFIER_LITERAL)));
   }
 
+  public SymbolicNameWrapper SYMBOLIC_NAME_WRAPPER() {
+    return b.<SymbolicNameWrapper>nonterminal(BicepLexicalGrammar.SYMBOLIC_NAME).is(
+      f.symbolicNameWrapper(
+        b.token(BicepLexicalGrammar.IDENTIFIER_LITERAL)));
```
Comment (nils-werner-sonarsource) [inline iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/BicepGrammar.java:762]:
```suggestion
        IDENTIFIER());
```

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/checks/SecureParameterDefaultValueCheck.java [hunk 2]
```diff
@@ -74,20 +73,10 @@ private static Predicate<Expression> isSensitiveDefaultValue() {
...
-        if (parameterName != null) {
-          return isSecureParameter(parameterName, parameterName.value());
-        }
-      }
-      return false;
-    };
+    // FixMe: should be checked for Parameter, but in Bicep parameters are Variables for now
+    return expr -> expr instanceof HasIdentifier hasIdentifier &&
+      hasIdentifier.identifier() instanceof Identifier identifier &&
+      isSecureParameter(identifier, identifier.value());
```
Comment (nils-werner-sonarsource) [inline iac-extensions/arm/src/main/java/org/sonar/iac/arm/checks/SecureParameterDefaultValueCheck.java:79]: How do you handle the case `parameter(unknown_expression)`? I think we will raise an FP at the moment. If yes, I think we should create a ticket to handle this.

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/checks/utils/CheckUtils.java [hunk 3]
```diff
@@ -141,4 +142,12 @@ public static StringLiteral parameterName(FunctionCall functionCall) {
     }
     return null;
   }
+
+  private static Expression maybeUnwrap(Expression expr) {
```
Comment (nils-werner-sonarsource) [inline iac-extensions/arm/src/main/java/org/sonar/iac/arm/checks/utils/CheckUtils.java:146]: I am not a fan of the words `unwrap` and `maybe` in function names. My suggestion would be `retrieveIdentifierOrExpression`. WDYT?

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/BicepGrammar.java [hunk 4]
```diff
@@ -688,8 +690,8 @@ public MultilineString MULTILINE_STRING() {
         b.token(Punctuator.TRIPLE_APOSTROPHE)));
   }
 
-  public FunctionCall FUNCTION_CALL() {
-    return b.<FunctionCall>nonterminal(BicepLexicalGrammar.FUNCTION_CALL).is(
+  public Expression FUNCTION_CALL() {
```
Comment (nils-werner-sonarsource) [inline iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/BicepGrammar.java:693]: Okay, for now, but the logic is now entirely in the Factory. We could also move it to the grammar. However, let's do it in a refactoring.

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/TreeFactory.java [hunk 5]
```diff
@@ -277,8 +281,19 @@ public ResourceDeclaration resourceDeclaration(
...
+    if ("parameters".equals(identifier.value()) || "variables".equals(identifier.value())) {
+      var isParameter = "parameters".equals(identifier.value());
+      var symbolicName = argumentList.get().elements().get(0);
+      var functionCallTextRange = TextRanges.merge(List.of(identifier.textRange(), rightParenthesis.textRange()));
+      if (isParameter) {
+        return new ParameterImpl(symbolicName, functionCallTextRange);
+      } else {
+        return new VariableImpl(symbolicName, functionCallTextRange);
+      }
+    }
```
Comment (nils-werner-sonarsource) [inline iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/TreeFactory.java:295]:
```suggestion
        var id = identifier.value();
    if ("parameters".equals(id) || "variables".equals(id)) {
      var symbolicName = argumentList.get().elements().get(0);
      var functionCallTextRange = TextRanges.merge(List.of(identifier.textRange(), rightParenthesis.textRange()));
      if ("parameters".equals(id)) {
        return new ParameterImpl(symbolicName, functionCallTextRange);
      } else {
        return new VariableImpl(symbolicName, functionCallTextRange);
      }
    }
```

File: iac-extensions/arm/src/test/java/org/sonar/iac/arm/tree/api/ExpressionAssert.java [hunk 6]
```diff
@@ -65,6 +66,13 @@ public IdentifierAssert asIdentifier() {
     return IdentifierAssert.assertThat((Identifier) actual);
   }
 
+  public IdentifierAssert asWrappedIdentifier() {
```
Comment (nils-werner-sonarsource) [inline iac-extensions/arm/src/test/java/org/sonar/iac/arm/tree/api/ExpressionAssert.java:69]: Also not a fan of `wrapped` but no better suggestion.

**Review by nils-werner-sonarsource — APPROVED** [review]
LGTM! Added some suggestions for renaming and refactoring.

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/checks/SecureParameterDefaultValueCheck.java (hunk 2 above)
Comment (petertrr) [reply iac-extensions/arm/src/main/java/org/sonar/iac/arm/checks/SecureParameterDefaultValueCheck.java:79]: Do you mean the case when `unknown_epxression` evaluates to a secure parameter name? Yes, that would be a FP. I'll create a ticket

**Pull request merged**